### DISCUSSION TITLE
Interact with JS smart-contract

### DISCUSSION THREAD

**Discussion by ChaoticTempest** [near/near-cli#978] [General]
## Deploy
This will deploy the contract to JS enclave:
```sh
near js deploy --accountId <my-account> --base64File <path> --deposit 0.1
```

## DevDeploy
Additionally, there is dev-deploy that doesn't require having an account already:
```sh
near js dev-deploy --base64File <path> --deposit 0.1
```

## Call
Call into a contract's call method:

```sh
near js call <contract-id> <function-name> --accountId <my-account> --args <args> --deposit 0.1
```

## View
Call into a contract's view method:

```sh
near js view <contract-id> <function-name> --accountId <my-account>
```

## Remove
This will remove the contract from the JS enclave:

```sh
near js delete --accountId <my-account>
```

## Feedback
As always, let us know where we can improve

**Comment by BenKurrek** [comment]
@ChaoticTempest why do we need the account ID flag for the view method?

**Reply by ChaoticTempest** [reply]
Yeah this was definitely not needed. This should fix it https://github.com/near/near-cli/pull/983

**Reply by BenKurrek** [reply]
Ah ok - approved the PR. Thanks!

**Comment by BenKurrek** [comment]
It looks like when u use `near js call`, you still need to pass in the arguments in an array: 
```bash
near js call $JS_CONTRACT set_greeting '["GO TEAM!"]' --accountId $JS_CONTRACT --deposit 0.1
```

If you try to use the regular way of interacting with the contracts (key value paired), it will throw:

```bash
near js call $JS_CONTRACT set_greeting '{"message": "GO TEAM!"}' --accountId $JS_CONTRACT --deposit 0.1
```

which throws:
```
ServerTransactionError: {"index":0,"kind":{"ExecutionError":"Smart contract panicked: value is not iterable\n    at set_greeting (build/contract.js:125)\n"}}
```

**Reply by ChaoticTempest** [reply]
Think full JSON deserialization isn't being done right now

cc @ailisp @volovyk-s

**Reply by austinbaggio** [reply]
@ailisp @volovyk-s @ChaoticTempest is this a large effort?

**Reply by ailisp** [reply]
It wasn't possible with old JS decorator approach, since a method object only knows it's a function object, nothing about the param name is saved. So we could only pass in argument by order. It should be possible with in current AST transformation approach with some effort, possibly a day:

https://esprima.org/demo/parse.html?code=class%20SomeContract%20%7B%0A%20%20%20%20%40call%0A%20%20%20%20someMethod(name%2C%20count)%20%7B%7D%0A%7D

Do we decide to do now? It is a breaking change and need modify all examples/demos/tests

**Reply by ChaoticTempest** [reply]
If it's possible, I think its best we do it now or at least before 1.0 is released, so long as there aren't any disadvantages here. If there aren't, it's better to be consistent with regular contract calls and how they do deserialization, since it would be strange if we started to differ too much from AS and RS contract calls
